File: src/main/cpp/subsystems/SwerveModule.cpp
#include "subsystems/SwerveModule.h"
#include <units/angle.h>
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/RobotBase.h>

/**
 * Constructor for the SwerveModule class
 * 
 * @param moduleID String for module identification
 * @param driveMotorChannel CAN ID of the drive Falcon
 * @param turningMotorChannel CAN ID of the turning Falcon
 * @param turningEncoderChannel CAN ID of turning ABS encoder
 * @param config The configuration data for this specific module
 */
SwerveModule::SwerveModule(std::string moduleID, int driveMotorChannel, int turningMotorChannel, int turningEncoderChannel, SwerveModuleConfig config)
    : moduleID(moduleID),
      m_driveMotor(driveMotorChannel),
      m_turningMotor(turningMotorChannel),
      m_turnEncoder(turningEncoderChannel),
      m_driveFeedforward{config.driveFf.GetFeedForward()},
      m_turnFeedforward{config.turnFf.GetFeedForward()},
      angle_offset(config.angleOffset),
      config(config)
{}

/**
 * Returns the current state of the module
 * 
 * @return frc::SwerveModuleState The current state of the swerve module
 */
frc::SwerveModuleState SwerveModule::GetState()
{
  currentState.angle = GetAngle();
  currentState.speed = GetVelocity();
  return currentState;
}

/**
 * Returns the current velocity of the module
 * 
 * @return units::meters_per_second_t The current velocity of the swerve module
 */
units::meters_per_second_t SwerveModule::GetVelocity()
{
  // Real Hardware
  return m_driveMotor.GetSelectedSensorVelocity(0) * 1_tr / 2048 * kDriveScaleFactor / 100_ms;
}

/**
 * Returns the current angle of the module
 * 
 * @return frc::Rotation2d The current angle of the swerve module
 */
frc::Rotation2d SwerveModule::GetAngle()
{
  // Real Hardware
  auto un_normalized = frc::Rotation2d(units::degree_t(m_turnEncoder.GetAbsolutePosition()));
  return frc::Rotation2d(un_normalized.Cos(), un_normalized.Sin());
}

/**
 * Returns the current angle of the module
 * 
 * @return frc::Rotation2d The current angle of the swerve module
 */
frc::Rotation2d SwerveModule::GetMotorAngle()
{
  auto pos_at_motor = frc::Rotation2d((units::radian_t) (m_turningMotor.GetSelectedSensorPosition() / 2048 / kTurnGearboxRatio * 1_tr));

  return frc::Rotation2d(pos_at_motor.Cos(), pos_at_motor.Sin());
}

void SwerveModule::Drive(units::meters_per_second_t target, bool openLoop)
{
  m_driveVolts = m_driveFeedforward.Calculate(target);

  auto target_in_native_units = target / kDriveScaleFactor * 100_ms * 2048 / 1_tr;

  if (openLoop) {
    m_driveMotor.SetVoltage(m_driveVolts);
  } else {
    m_driveMotor.Set(ControlMode::Velocity, target_in_native_units, DemandType::DemandType_ArbitraryFeedForward, m_driveVolts / 12_V);
  }
}

void SwerveModule::Turn(frc::Rotation2d target)
{
  auto target_in_native_units = target.Radians() / 1_tr * kTurnGearboxRatio * 2048;  

  m_turningMotor.Set(TalonFXControlMode::MotionMagic, target_in_native_units);
}

/**
 * Sets the module's target state
 * 
 * @param state The module's target state
 */
void SwerveModule::SetModule(const frc::SwerveModuleState &state, bool openLoop)
{
  currentState = GetState();

  targetState = frc::SwerveModuleState::Optimize(state, currentState.angle);

  Drive(targetState.speed, openLoop);

  Turn(targetState.angle.Radians());
}

/**
 * Just Stop....
 *
 */
void SwerveModule::Stop()
{
  m_driveMotor.SetVoltage(0_V);
  m_turningMotor.SetVoltage(0_V);

  // Simulation
  m_driveSim.SetInputVoltage(0_V);
  m_turnSim.SetInputVoltage(0_V);
}

void SwerveModule::ConfigureSystem()
{
  if (frc::RobotBase::IsSimulation())
  {
    m_driveFeedforward.kS = 0_V;
    m_turnFeedforward.kS = 0_V;
  }

  // Drive Motor Configuration
  m_driveMotor.ConfigFactoryDefault();
  m_driveMotor.SetInverted(true); // Remember: forward-positive!
  m_driveMotor.SetNeutralMode(ctre::phoenix::motorcontrol::NeutralMode::Brake);
  m_driveMotor.ConfigSupplyCurrentLimit(SupplyCurrentLimitConfiguration(true, kDriveMotorCurrentLimit.value(), kDriveMotorCurrentLimit.value(), 0.0));
  m_driveMotor.SetSensorPhase(false);

  {
    auto allSettings = TalonFXConfiguration();
    m_driveMotor.GetAllConfigs(allSettings);

    allSettings.supplyCurrLimit = {true, kDriveMotorCurrentLimit / 1_A, kDriveMotorCurrentLimit / 1_A, 0.0};
    allSettings.slot0.kP = config.drivePID.kP;
    allSettings.slot0.kD = config.drivePID.kD;

    auto accel_in_drive_motor_frame = config.drivePID.constraints.maxVelocity / kDriveScaleFactor;
    auto accel_in_drive_motor_units = accel_in_drive_motor_frame * 2048 / 1_tr * 100_ms * 1_s;

    allSettings.motionAcceleration = accel_in_drive_motor_units;

    m_driveMotor.ConfigAllSettings(allSettings);
  }


  // Turning Motor Configuration
  m_turningMotor.ConfigFactoryDefault();
  m_turningMotor.SetInverted(false); // Remember: forward-positive!
  m_turningMotor.SetNeutralMode(ctre::phoenix::motorcontrol::NeutralMode::Brake);

  {
    auto allSettings = TalonFXConfiguration();
    m_turningMotor.GetAllConfigs(allSettings);

    allSettings.supplyCurrLimit = {true, kTurningMotorCurrentLimit / 1_A, kTurningMotorCurrentLimit / 1_A, 0.0};
    allSettings.slot0.kP = config.turnPID.kP;
    allSettings.slot0.kD = config.turnPID.kD;

    auto vel_in_turn_motor_frame = config.turnPID.constraints.maxVelocity * kTurnGearboxRatio;
    auto vel_in_turn_motor_units = vel_in_turn_motor_frame * 2048 / 1_tr * 100_ms;

    auto accel_in_turn_motor_frame = config.turnPID.constraints.maxAcceleration * kTurnGearboxRatio;
    auto accel_in_turn_motor_units = accel_in_turn_motor_frame * 2048 / 1_tr * 100_ms * 1_s;

    allSettings.motionCruiseVelocity = vel_in_turn_motor_units;
    allSettings.motionAcceleration = accel_in_turn_motor_units;

    m_turningMotor.ConfigAllSettings(allSettings);
  }

  // Turning Encoder Config
  ctre::phoenix::sensors::CANCoderConfiguration encoderConfig;
  m_turnEncoder.GetAllConfigs(encoderConfig);
  encoderConfig.enableOptimizations = true;
  encoderConfig.initializationStrategy = ctre::phoenix::sensors::SensorInitializationStrategy::BootToAbsolutePosition;
  encoderConfig.absoluteSensorRange = ctre::phoenix::sensors::AbsoluteSensorRange::Signed_PlusMinus180;
  // encoderConfig.magnetOffsetDegrees = angle_offset / 1_deg;
  encoderConfig.sensorDirection = false;
  m_turnEncoder.ConfigAllSettings(encoderConfig);
}

void SwerveModule::RegisterDataEntries(wpi::log::DataLog &log)
{
}

void SwerveModule::LogDataEntries(wpi::log::DataLog &log)
{

}

void SwerveModule::UpdateTelemetry()
{
  
}

void SwerveModule::InitSendable(wpi::SendableBuilder &builder)
{
  builder.SetSmartDashboardType("SwerveModule");
  builder.SetActuator(true);

  // Turning Encoders
  builder.AddDoubleProperty(
      moduleID + "/abs_encoder", [this] { return GetAngle().Radians() / 1_rad; }, nullptr);
  builder.AddDoubleProperty(
      moduleID + "/turn_motor_encoder", [this] { return GetMotorAngle().Radians() / 1_rad; }, nullptr);

  builder.AddDoubleProperty(moduleID + "/raw_drive_volts", [this] { return m_driveVolts / 1_V; }, nullptr);
  builder.AddDoubleProperty(moduleID + "/sim_drive_volts", [this] { return driveMotorSim.GetMotorOutputLeadVoltage(); }, nullptr);
  builder.AddDoubleProperty(moduleID + "/raw_turn_volts", [this] { return m_turnVolts / 1_V; }, nullptr);
  builder.AddDoubleProperty(moduleID + "/sim_turn_volts", [this] { return turningMotorSim.GetMotorOutputLeadVoltage(); }, nullptr);

  // Thermal
  // builder.AddDoubleProperty(
  //     moduleID + "/Drive Temp [C]", [this] { return m_driveMotor.GetTemperature(); }, nullptr);
  // builder.AddDoubleProperty(
  //     moduleID + "/Angle Temp [C]", [this] { return m_turningMotor.GetTemperature(); }, nullptr);

  builder.AddDoubleProperty(
    moduleID + "/state/speed", [this] {return GetState().speed / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
    moduleID + "/state/angle", [this] {return GetState().angle.Radians() / 1_rad; }, nullptr);

  builder.AddDoubleProperty(
    moduleID + "/goal/speed", [this] {return targetState.speed / 1_mps; }, nullptr);
  builder.AddDoubleProperty(
    moduleID + "/goal/angle", [this] {return targetState.angle.Radians() / 1_rad; }, nullptr);

  builder.AddDoubleProperty(moduleID + "/sim/speed", [this] { return m_driveSim.GetLinearVelocity() / 1_mps; }, nullptr);
  builder.AddDoubleProperty(moduleID + "/sim/angular_speed", [this] { return m_driveSim.GetAngularVelocity() / 1_rad_per_s; }, nullptr);
  builder.AddDoubleProperty(moduleID + "/sim/angle", [this] { return m_turnSim.GetAngle() / 1_rad; }, nullptr);

  builder.AddDoubleProperty(moduleID + "/drive/PID/kP", [this] {
    return driveSlotConfig.kP;
  },
  [this] (double val) {
    driveSlotConfig.kP = val;
    m_driveMotor.ConfigureSlot(driveSlotConfig, 0, 0);
  });
  builder.AddDoubleProperty(moduleID + "/drive/PID/kD", [this] {
    return driveSlotConfig.kD;
  },
  [this] (double val) {
    driveSlotConfig.kD = val;
    m_driveMotor.ConfigureSlot(driveSlotConfig, 0, 0);
  });

    builder.AddDoubleProperty(moduleID + "/turn/PID/kP", [this] {
    return turnSlotConfig.kP;
  },
  [this] (double val) {
    turnSlotConfig.kP = val;
    m_turningMotor.ConfigureSlot(turnSlotConfig, 0, 0);
  });
  builder.AddDoubleProperty(moduleID + "/turn/PID/kD", [this] {
    return turnSlotConfig.kD;
  },
  [this] (double val) {
    turnSlotConfig.kD = val;
    m_turningMotor.ConfigureSlot(turnSlotConfig, 0, 0);
  });
}

units::ampere_t SwerveModule::SimDrive(units::volt_t battery)
{
  driveMotorSim.SetBusVoltage(12); //battery / 1_V);
  // -1 because drive motor is inverted
  m_driveSim.SetInputVoltage(-1_V * driveMotorSim.GetMotorOutputLeadVoltage());

  m_driveSim.Update(20_ms);

  driveMotorSim.SetStatorCurrent(m_driveSim.GetCurrentDraw() / 1_A);

  auto rpm_at_wheel = m_driveSim.GetAngularVelocity();
  auto rpm_at_motor = rpm_at_wheel * kDriveGearboxRatio;
  auto rpm_in_native_units = rpm_at_motor / 1_tr * 2048 * 100_ms;

  driveMotorSim.SetIntegratedSensorVelocity(rpm_in_native_units);

  return m_driveSim.GetCurrentDraw();
}

units::ampere_t SwerveModule::SimTurn(units::volt_t battery)
{

  turningMotorSim.SetBusVoltage(12); //battery / 1_V);
  encoderSim.SetBusVoltage(12); //battery / 1_V);
  m_turnSim.SetInputVoltage(1_V * turningMotorSim.GetMotorOutputLeadVoltage());

  m_turnSim.Update(20_ms);

  turningMotorSim.SetStatorCurrent(m_turnSim.GetCurrentDraw() / 1_A);

  auto angular_pos = m_turnSim.GetAngle();
  auto angular_vel = m_turnSim.GetAngularVelocity();

  // native cancoder units are 4096 ticks per rotation, vel measured per 100ms
  auto angular_pos_at_cancoder = angular_pos / 1_tr * 4096;
  auto angular_vel_at_cancoder = angular_vel / 1_tr * 4096 * 100_ms;

  encoderSim.SetRawPosition(angular_pos_at_cancoder);
  encoderSim.SetVelocity(angular_vel_at_cancoder);

  // native falcon units are 2048 ticks per rotation, vel measured per 100ms
  auto angular_pos_at_motor = angular_pos / 1_tr * kTurnGearboxRatio * 2048;
  auto angular_vel_at_motor = angular_vel / 1_tr * kTurnGearboxRatio * 2048 * 100_ms;

  turningMotorSim.SetIntegratedSensorRawPosition(angular_pos_at_motor);
  turningMotorSim.SetIntegratedSensorVelocity(angular_vel_at_motor);

  return m_turnSim.GetCurrentDraw();
}

units::ampere_t SwerveModule::SimPeriodic(units::volt_t battery)
{
  return SimDrive(battery) + SimTurn(battery);
}

void SwerveModule::SimInit()
{
  
}

ErrorCode SwerveModule::SeedTurnMotor()
{
  auto position_at_cancoder = m_turnEncoder.GetAbsolutePosition() * 1_deg;
  auto position_at_motor = position_at_cancoder / 1_tr * kTurnGearboxRatio * 2048;

  auto error = m_turnEncoder.GetLastError();
  if (error != ErrorCode::OK) {
    return error;
  }

  return m_turningMotor.SetSelectedSensorPosition(position_at_motor);
}

units::radians_per_second_t SwerveModule::AngularVelocity()
{
  auto motor_rps = m_turningMotor.GetSelectedSensorVelocity() * 1_tr / 2048 / 100_ms;

  return motor_rps / kTurnGearboxRatio;
}

File: src/main/java/io/robojackets/subsystems/SwerveModule.java
package io.robojackets.subsystems;

import com.ctre.phoenix.ErrorCode;
import com.ctre.phoenix.motorcontrol.DemandType;
import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.SupplyCurrentLimitConfiguration;
import com.ctre.phoenix.motorcontrol.TalonFXControlMode;
import com.ctre.phoenix.motorcontrol.TalonFXInvertType;
import com.ctre.phoenix.motorcontrol.TalonFXSimCollection;
import com.ctre.phoenix.motorcontrol.can.TalonFXConfiguration;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
import com.ctre.phoenix.sensors.AbsoluteSensorRange;
import com.ctre.phoenix.sensors.CANCoder;
import com.ctre.phoenix.sensors.CANCoderConfiguration;
import com.ctre.phoenix.sensors.CANCoderSimCollection;
import com.ctre.phoenix.sensors.SensorInitializationStrategy;
import edu.wpi.first.math.Nat;
import edu.wpi.first.math.numbers.N1;
import edu.wpi.first.math.numbers.N2;
import edu.wpi.first.util.datalog.DataLog;
import edu.wpi.first.util.sendable.SendableBuilder;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.y2023.math.Matrix;
import edu.wpi.first.y2023.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.y2023.math.geometry.Rotation2d;
import edu.wpi.first.y2023.math.kinematics.SwerveModuleState;
import edu.wpi.first.y2023.math.system.LinearSystem;
import edu.wpi.first.y2023.math.system.plant.DCMotor;
import edu.wpi.first.y2023.math.system.plant.LinearSystemId;
import edu.wpi.first.y2023.math.util.Units;
import io.robojackets.config.SwerveModuleConfig;
import io.robojackets.sim.SwerveWheelSim;
import io.robojackets.sim.TurretSim;

public class SwerveModule extends Subsystem {
  private static final double WHEEL_RADIUS_METERS = Units.inchesToMeters(2);
  private static final double DRIVE_GEARBOX_RATIO = 6.75;
  private static final double TURN_GEARBOX_RATIO = 12.8;
  private static final double DRIVE_SCALE_FACTOR = WHEEL_RADIUS_METERS / DRIVE_GEARBOX_RATIO;

  private static final double DRIVE_CURRENT_LIMIT_AMPS = 40;
  private static final double TURN_CURRENT_LIMIT_AMPS = 30;

  private final SimpleMotorFeedforward driveFeedForward;
  private final SimpleMotorFeedforward turnFeedForward;

  private final double angleOffset;

  private SwerveModuleState currentState = new SwerveModuleState();
  private SwerveModuleState targetState = new SwerveModuleState();

  private final String moduleID;

  private final SwerveModuleConfig config;

  private final WPI_TalonFX driveMotor;
  private final WPI_TalonFX turnMotor;
  private final CANCoder turnEncoder;

  private double driveVolts;
  private double turnVolts;

  private final TalonFXSimCollection driveMotorSim;
  private final TalonFXSimCollection turnMotorSim;
  private final CANCoderSimCollection turnEncoderSim;

  private final LinearSystem<N1, N1, N1> drivePlant;

  private final SwerveWheelSim driveSim;

  private final LinearSystem<N2, N1, N1> turnPlant;

  private final TurretSim turnSim;

  public SwerveModule(
      String moduleID, int driveID, int turnID, int encID, SwerveModuleConfig config) {
    this.moduleID = moduleID;
    driveMotor = new WPI_TalonFX(driveID);
    turnMotor = new WPI_TalonFX(turnID);
    turnEncoder = new CANCoder(encID);

    this.config = config;

    angleOffset = config.getAngleOffsetRadians();

    driveFeedForward =
        new SimpleMotorFeedforward(
            RobotBase.isSimulation() ? 0 : config.getDriveFeedForwardConfig().getKS(),
            config.getDriveFeedForwardConfig().getKV(),
            config.getDriveFeedForwardConfig().getKA());
    turnFeedForward =
        new SimpleMotorFeedforward(
            RobotBase.isSimulation() ? 0 : config.getTurnFeedForwardConfig().getKS(),
            config.getTurnFeedForwardConfig().getKV(),
            config.getTurnFeedForwardConfig().getKA());

    driveMotorSim = driveMotor.getSimCollection();
    turnMotorSim = turnMotor.getSimCollection();
    turnEncoderSim = turnEncoder.getSimCollection();

    drivePlant =
        LinearSystemId.identifyVelocitySystem(
            config.getDriveFeedForwardConfig().getKV() * WHEEL_RADIUS_METERS,
            config.getDriveFeedForwardConfig().getKA() * WHEEL_RADIUS_METERS);

    driveSim =
        new SwerveWheelSim(
            drivePlant,
            DCMotor.getFalcon500(1),
            DRIVE_GEARBOX_RATIO,
            WHEEL_RADIUS_METERS,
            Matrix.mat(Nat.N1(), Nat.N1()).fill(0.0));

    turnPlant =
        LinearSystemId.identifyPositionSystem(
            config.getTurnFeedForwardConfig().getKV(), config.getTurnFeedForwardConfig().getKA());

    turnSim =
        new TurretSim(
            turnPlant,
            DCMotor.getFalcon500(1),
            TURN_GEARBOX_RATIO,
            Matrix.mat(Nat.N1(), Nat.N1()).fill(1.0));
  }

  public SwerveModuleState GetState() {
    return new SwerveModuleState(GetVelocity(), GetAngle());
  }

  public double GetVelocity() {
    return driveMotor.getSelectedSensorVelocity() / 2048 * DRIVE_SCALE_FACTOR / 0.100;
  }

  public Rotation2d GetAngle() {
    var un_normalized = Rotation2d.fromDegrees(turnEncoder.getAbsolutePosition());
    return new Rotation2d(un_normalized.getCos(), un_normalized.getSin());
  }

  public Rotation2d GetMotorAngle() {
    return Rotation2d.fromRotations(
        turnMotor.getSelectedSensorPosition() / 2048 / TURN_GEARBOX_RATIO);
  }

  void Drive(double targetMetersPerSecond, boolean openLoop) {
    driveVolts = driveFeedForward.calculate(targetMetersPerSecond);

    double target_in_native_units = targetMetersPerSecond / DRIVE_SCALE_FACTOR * 0.100 * 2048;

    if (openLoop) {
      driveMotor.setVoltage(driveVolts);
    } else {
      driveMotor.set(
          TalonFXControlMode.Velocity,
          target_in_native_units,
          DemandType.ArbitraryFeedForward,
          driveVolts / 12);
    }
  }

  void Turn(Rotation2d target) {
    Rotation2d normalized_target_angle = new Rotation2d(target.getCos(), target.getSin());
    Rotation2d normalized_current_angle = GetState().angle;
    Rotation2d actual_current_angle = GetMotorAngle();

    Rotation2d target_diff = normalized_target_angle.minus(normalized_current_angle);
    double actual_target_angle = actual_current_angle.getRadians() + target_diff.getRadians();

    double target_in_native_units =
        Units.radiansToRotations(actual_target_angle) * TURN_GEARBOX_RATIO * 2048;

    turnMotor.set(TalonFXControlMode.MotionMagic, target_in_native_units);
  }

  void SetModule(final SwerveModuleState state, boolean openLoop) {
    currentState = GetState();
    targetState = SwerveModuleState.optimize(state, currentState.angle);

    Drive(targetState.speedMetersPerSecond, openLoop);
    Turn(targetState.angle);
  }

  void Stop() {
    driveMotor.stopMotor();
    turnMotor.stopMotor();

    driveSim.setInputVoltage(0);
    turnSim.setInputVoltage(0);
  }

  @Override
  public void initSendable(SendableBuilder builder) {

    builder.setSmartDashboardType("SwerveModule");
    builder.setActuator(true);

    // Turning Encoders
    builder.addDoubleProperty(moduleID + "/abs_encoder", () -> GetAngle().getRadians(), null);
    builder.addDoubleProperty(
        moduleID + "/turn_motor_encoder", () -> GetMotorAngle().getRadians(), null);

    builder.addDoubleProperty(moduleID + "/raw_drive_volts", () -> driveVolts, null);
    builder.addDoubleProperty(
        moduleID + "/sim_drive_volts", () -> driveMotorSim.getMotorOutputLeadVoltage(), null);
    builder.addDoubleProperty(moduleID + "/raw_turn_volts", () -> turnVolts, null);
    builder.addDoubleProperty(
        moduleID + "/sim_turn_volts", () -> turnMotorSim.getMotorOutputLeadVoltage(), null);

    // Thermal
    builder.addDoubleProperty(
        moduleID + "/Drive Temp [C]", () -> driveMotor.getTemperature(), null);
    builder.addDoubleProperty(moduleID + "/Angle Temp [C]", () -> turnMotor.getTemperature(), null);

    builder.addDoubleProperty(
        moduleID + "/state/speed", () -> GetState().speedMetersPerSecond, null);
    builder.addDoubleProperty(moduleID + "/state/angle", () -> GetState().angle.getRadians(), null);

    builder.addDoubleProperty(
        moduleID + "/goal/speed", () -> targetState.speedMetersPerSecond, null);
    builder.addDoubleProperty(moduleID + "/goal/angle", () -> targetState.angle.getRadians(), null);

    builder.addDoubleProperty(moduleID + "/sim/speed", () -> driveSim.GetLinearVelocity(), null);
    builder.addDoubleProperty(
        moduleID + "/sim/turret_speed", () -> turnSim.getAngularVelocityRadiansPerSecond(), null);
    builder.addDoubleProperty(moduleID + "/sim/angle", () -> turnSim.getAngleRadians(), null);

    builder.addDoubleProperty(
        moduleID + "/turn/SensorVel", () -> turnMotor.getSelectedSensorVelocity(0), null);
    builder.addDoubleProperty(
        moduleID + "/turn/SensorPos", () -> turnMotor.getSelectedSensorPosition(0), null);
    builder.addDoubleProperty(
        moduleID + "/turn/MotorOutputPercent", () -> turnMotor.getMotorOutputPercent(), null);
    builder.addDoubleProperty(
        moduleID + "/turn/MotorOutputSimVoltage",
        () -> turnMotorSim.getMotorOutputLeadVoltage(),
        null);
    builder.addDoubleProperty(
        moduleID + "/turn/ClosedLoopError", () -> turnMotor.getClosedLoopError(0), null);
    builder.addDoubleProperty(
        moduleID + "/turn/ClosedLoopTarget", () -> turnMotor.getClosedLoopTarget(0), null);

    builder.addDoubleProperty(
        moduleID + "/drive/SensorVel", () -> driveMotor.getSelectedSensorVelocity(0), null);
    builder.addDoubleProperty(
        moduleID + "/drive/SensorPos", () -> driveMotor.getSelectedSensorPosition(0), null);
    builder.addDoubleProperty(
        moduleID + "/drive/MotorOutputPercent", () -> driveMotor.getMotorOutputPercent(), null);
    builder.addDoubleProperty(
        moduleID + "/drive/MotorOutputSimVoltage",
        () -> driveMotorSim.getMotorOutputLeadVoltage(),
        null);
    builder.addDoubleProperty(
        moduleID + "/drive/ClosedLoopError", () -> driveMotor.getClosedLoopError(0), null);
    builder.addDoubleProperty(
        moduleID + "/drive/ClosedLoopTarget", () -> driveMotor.getClosedLoopTarget(0), null);
  }

  @Override
  public void ConfigureSystem() {
    driveMotor.setSensorPhase(false);
    driveMotor.setInverted(TalonFXInvertType.Clockwise);
    driveMotor.setNeutralMode(NeutralMode.Brake);
    driveMotor.configFactoryDefault();

    TalonFXConfiguration driveSettings = new TalonFXConfiguration();
    driveMotor.getAllConfigs(driveSettings);

    driveSettings.supplyCurrLimit =
        new SupplyCurrentLimitConfiguration(
            true, DRIVE_CURRENT_LIMIT_AMPS, DRIVE_CURRENT_LIMIT_AMPS, 0);
    driveSettings.slot0.kP = config.getDrivePIDConfig().getKP();
    driveSettings.slot0.kD = config.getDrivePIDConfig().getKD();

    driveMotor.configAllSettings(driveSettings);

    turnMotor.setSensorPhase(false);
    turnMotor.setInverted(TalonFXInvertType.CounterClockwise);
    turnMotor.setNeutralMode(NeutralMode.Brake);
    turnMotor.configFactoryDefault();

    TalonFXConfiguration turnSettings = new TalonFXConfiguration();
    turnMotor.getAllConfigs(turnSettings);

    turnSettings.supplyCurrLimit =
        new SupplyCurrentLimitConfiguration(
            true, TURN_CURRENT_LIMIT_AMPS, TURN_CURRENT_LIMIT_AMPS, 0);
    turnSettings.slot0.kP = config.getTurnPIDConfig().getKP();

    double vel_in_turn_motor_frame =
        config.getMaxTurnVelocityRadiansPerSecond() * TURN_GEARBOX_RATIO;
    double vel_in_turn_motor_units =
        Units.radiansToRotations(vel_in_turn_motor_frame) * 2048 * 0.100;

    double accel_in_turn_motor_frame =
        config.getMaxTurnAccelRadiansPerSecondPerSecond() * TURN_GEARBOX_RATIO;
    double accel_in_turn_motor_units =
        Units.radiansToRotations(accel_in_turn_motor_frame) * 2048 * 0.100;

    turnSettings.motionCruiseVelocity = vel_in_turn_motor_units;
    turnSettings.motionAcceleration = accel_in_turn_motor_units;

    turnMotor.configAllSettings(turnSettings);

    CANCoderConfiguration encoderSettings = new CANCoderConfiguration();
    turnEncoder.getAllConfigs(encoderSettings);

    encoderSettings.initializationStrategy = SensorInitializationStrategy.BootToAbsolutePosition;
    encoderSettings.absoluteSensorRange = AbsoluteSensorRange.Signed_PlusMinus180;
    if (RobotBase.isReal()) {
      encoderSettings.magnetOffsetDegrees = Units.radiansToDegrees(angleOffset);
    }

    encoderSettings.sensorDirection = false;
    turnEncoder.configAllSettings(encoderSettings);
  }

  @Override
  public void RegisterDataEntries(DataLog log) {
    // TODO Auto-generated method stub

  }

  @Override
  public void LogDataEntries(DataLog log) {
    // TODO Auto-generated method stub

  }

  @Override
  public void SimInit() {
    // TODO Auto-generated method stub

  }

  @Override
  public void SimPeriodic(double battery, double[] currentDraw) {
    // TODO Auto-generated method stub
    double driveAmps = SimDrive(battery);
    double turnAmps = SimTurn(battery);
  }

  private double SimDrive(double battery) {
    driveMotorSim.setBusVoltage(12);
    driveSim.setInputVoltage(-driveMotorSim.getMotorOutputLeadVoltage());

    driveSim.update(0.02);

    driveMotorSim.setStatorCurrent(driveSim.getCurrentDrawAmps());

    double speed_at_wheel = driveSim.getAngularVelocityRadPerSec();
    double speed_at_motor = speed_at_wheel * DRIVE_GEARBOX_RATIO;
    double speed_in_native_units = Units.radiansToRotations(speed_at_motor) * 2048 * 0.100;

    driveMotorSim.setIntegratedSensorVelocity(-(int) speed_in_native_units);

    return driveSim.getCurrentDrawAmps();
  }

  private double SimTurn(double battery) {
    turnMotorSim.setBusVoltage(12);
    turnEncoderSim.setBusVoltage(12);

    turnSim.setInputVoltage(turnMotorSim.getMotorOutputLeadVoltage());

    double angular_pos = turnSim.getAngleRadians();
    double angular_vel = turnSim.getAngularVelocityRadiansPerSecond();

    double angular_pos_at_encoder = Units.radiansToRotations(angular_pos) * 4096;
    double angular_vel_at_encoder = Units.radiansToRotations(angular_vel) * 4096 * 0.100;

    turnEncoderSim.setRawPosition((int) angular_pos_at_encoder);
    turnEncoderSim.setVelocity((int) angular_vel_at_encoder);

    double angular_pos_at_motor = Units.radiansToRotations(angular_pos) * TURN_GEARBOX_RATIO * 2048;
    double angular_vel_at_motor =
        Units.radiansToRotations(angular_vel) * TURN_GEARBOX_RATIO * 2048 * 0.100;

    turnMotorSim.setIntegratedSensorRawPosition((int) angular_pos_at_motor);
    turnMotorSim.setIntegratedSensorVelocity((int) angular_vel_at_motor);

    return turnSim.getCurrentDrawAmps();
  }

  public ErrorCode SeedTurnMotor() {
    double pos_at_encoder = Units.degreesToRotations(turnEncoder.getAbsolutePosition());
    double pos_at_motor = pos_at_encoder * TURN_GEARBOX_RATIO * 2048;

    turnMotor.stopMotor();

    return turnMotor.setSelectedSensorPosition(pos_at_motor);
  }

  public double GetAngularVelocityRotationsPerSecond() {
    double motor_rps = turnMotor.getSelectedSensorVelocity() / 2048 / 0.100;

    return motor_rps / TURN_GEARBOX_RATIO;
  }
}
